_posts/2020-04-17-quarkus-mutual-tls.adoc

```diff
@@ -0,0 +1,405 @@
+
+=== Configure MicroProfile rest client for Mutual TLS.
+
+Add the following properties to enable SSL in your application `src/main/resources/application.properties`
+
+.application.properties
+[source]
+----
+org.acme.client.mtls.GreetingService/mp-rest/url=https://localhost:8443
+org.acme.client.mtls.GreetingService/mp-rest/trustStore=classpath:/META-INF/resources/client.truststore
+org.acme.client.mtls.GreetingService/mp-rest/trustStorePassword=password
+org.acme.client.mtls.GreetingService/mp-rest/keyStore=classpath:/META-INF/resources/client.keystore
+org.acme.client.mtls.GreetingService/mp-rest/keyStorePassword=password
+
+quarkus.ssl.native=true
+----
+
+=== Run it.
+
+[source]
+----
+mvn quarkus:dev
+----
+
+Now let's hit the client _/hello_ endpoint, and you should expect the following.
+
+[source]
+----
+curl http://localhost:8080/hello
+hello from server
+----
+
+== External Configuration
+
+You do not want to have a certificate inside your application and Quarkus allows you to use external configuration and override the runtime application properties.
+Using the external `config` directory under the current working directory, for instance `config/application.properties`
+
+Let's check how we can do in Kubernetes / OpenShift.
+
+[NOTE]
+====
+Working directory:
+
+* `redhat-openjdk-18/openjdk18-openshift` image is `/deployments`
+* `ubi-quarkus-native-s2i:19.3.1-java11` image is `/home/quarkus`
+====
+
+=== Secret
+
+Create server, client and truststore secret which contains your certificates and truststore. For instance:
+[source]
+----
+kubectl create secret generic server --from-file=tls/server/
+kubectl create secret generic client --from-file=tls/client/
+kubectl create secret generic truststore --from-file=tls/ca/truststore
+----
+
+=== ConfigMap
+
+Create the server and client ConfigMap.
+
+.server-ConfigMap.yaml
+[source,yaml]
+----
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: server
+data:
+  application.properties: |
+    quarkus.http.ssl.certificate.key-store-file=/deployments/tls/server.keystore
+    quarkus.http.ssl.certificate.key-store-password=password
+    quarkus.http.ssl.certificate.trust-store-file=/deployments/tls/ca/truststore
+    quarkus.http.ssl.certificate.trust-store-password=password
+----
+
+.client-ConfigMap.yaml
+[source,yaml]
+----
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: client
+data:
+  application.properties: |
+    org.acme.client.mtls.GreetingService/mp-rest/url=https://server:8443
+    org.acme.client.mtls.GreetingService/mp-rest/trustStore=/deployments/tls/ca/truststore
+    org.acme.client.mtls.GreetingService/mp-rest/trustStorePassword=password
+    org.acme.client.mtls.GreetingService/mp-rest/keyStore=/deployments/tls/client.keystore
+    org.acme.client.mtls.GreetingService/mp-rest/keyStorePassword=password
+----
+
+=== Mount Everything
+
+Mount everything in your Deployment Application. 
+Here an example for the client application:
+
+[source,yaml]
```

Comment:
@geoand do we have a better way to do this with the kube / openshift extensions? This is a lot of manual yaml to enter.

Reply:
Yes, the Kubernetes extension is configurable in a way that allows you to specify custom mounts: https://github.com/quarkusio/quarkus/blob/6106870c0fb6e4b69a92e702f4615ba15c43576b/extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/KubernetesConfig.java#L138

Furthermore, with the new `kubernetes-config` extension, you can now read ConfigMap contents **without** having to  mount them in the pod: https://github.com/quarkusio/quarkus/blob/6c124cfbfe8e596e2d1e1363f91a4b3342394360/docs/src/main/asciidoc/kubernetes-config.adoc

So ideally this section would be reworked to showcase that

Reply:
@geoand, @emmanuelbernard  you are right but sometimes you in a restricted environment which not allow the role to read/view ConfigMap / Secret to the service account. Any way I can adjust the quickstart to use the kubernetes-config, to show the capability

Reply:
Hi @geoand and @emmanuelbernard.
I added the kubernetes-config extensions scenario and as well more information around the manual mounting in case they do not have role permission. This should cover all the case.
And everything using the 1.4.0.Final release.
Let me know if you want to wait the 1.4.1.Final release.

Reply:
`1.4.1.Final` is right around the corner and that is version we will be telling people to use, so I think it makes sense to wait.

Reply:
Everything is ready on 1.4.1.Final.